File: ba-translate.py
#!/usr/bin/env python3
import json, os, textwrap, re

###==============================================
### Helper functions
def toEnumStr(string: str):
    return re.sub(r'([a-z](?=[A-Z])|[A-Z](?=[A-Z][a-z]))', r'\1_', string).lower()

def stripName(string: str):
    string = string.replace("_default", "")
    string = string.replace("_", " ")

    wrongNames = [
        ("Aris", "Alice"),
        ("Hihumi", "Hifumi"),
        ("Zunko", "Junko")
    ]
    for wrongName in wrongNames:
        string = string.replace(wrongName[0], wrongName[1])
    
    return string

###==============================================
### Schemes
def toZZZString(student):
    result = textwrap.dedent(f"""\
    id: {student["Id"]}
    dev name: {student["DevName"]}
    production step: {student["ProductionStep"]} 


    ### ------- INFO -------
    name: {student["StrippedName"]}
    rarity: {student["Rarity"]}{(" : " + '★'*student["DefaultStarGrade"] + '☆'*(student["MaxStarGrade"] - student["DefaultStarGrade"])) if student["MaxStarGrade"] > 0 else ""}

    school: {student["School"]}
    club: "{student["Club"]}"
    tags: {" :; ".join('"{}"'.format(tag) for tag in student["Tags"]) if len(student["Tags"]) > 0 else '""'}


    ### ------- STATUS -------
    type: {"striker" if student["SquadType"] == "Main" else "special" if student["SquadType"] == "Support" else student["SquadType"].lower()}
    role: {toEnumStr(student["TacticRole"])}
    position: {toEnumStr(student["TacticRange"])}

    armor: {toEnumStr(student["ArmorType"])}
    bullet: {toEnumStr(student["BulletType"])}

    weapon: {student["WeaponType"] if student["WeaponType"].isupper() else toEnumStr(student["WeaponType"])}
    aim ik: {toEnumStr(student["AimIKType"])}

    equipments: {" :; ".join([equipment.lower() for equipment in student["EquipmentSlot"]]) if len(student["EquipmentSlot"]) > 0 else "null"}
    """)

    if "stat" in student.keys():
        result = result + textwrap.dedent(f"""

        ### ------- STATS -------
        hp:
          min: {student["stat"]["MaxHP1"]}
          max: {student["stat"]["MaxHP100"]}
        atk:
          min: {student["stat"]["AttackPower1"]}
          max: {student["stat"]["AttackPower100"]}
        def:
          min: {student["stat"]["DefensePower1"]}
          max: {student["stat"]["DefensePower100"]}
        heal:
          min: {student["stat"]["HealPower1"]}
          max: {student["stat"]["HealPower100"]}

        stability:
          rate: {student["stat"]["StabilityRate"]} : {student["stat"]["StabilityRate"] / 100:.0f}%
          point: {student["stat"]["StabilityPoint"]}

        critical:
          crit point: {student["stat"]["CriticalPoint"]}
          crit dmg rate: {student["stat"]["CriticalDamageRate"]} : {student["stat"]["CriticalDamageRate"] / 100:.0f}%

          res point: {student["stat"]["CriticalResistPoint"]}
          res dmg rate: {student["stat"]["CriticalDamageResistRate"]} : {student["stat"]["CriticalDamageResistRate"] / 100:.0f}%

        oppression:
          power: {student["stat"]["OppressionPower"]}
          resist: {student["stat"]["OppressionResist"]}

        ammo:
          cost: {student["stat"]["AmmoCost"]}
          count: {student["stat"]["AmmoCount"]}

        range: {student["stat"]["Range"]}
        regen: {student["stat"]["RegenCost"]}
        dodge: {student["stat"]["DodgePoint"]}
        accuracy: {student["stat"]["AccuracyPoint"]}

        environment:
          street: {student["stat"]["StreetBattleAdaptation"]}
          indoor: {student["stat"]["IndoorBattleAdaptation"]}
          outdoor: {student["stat"]["OutdoorBattleAdaptation"]}
        """)
    
    if "profile" in student.keys():
        result = result + textwrap.dedent(f"""

        ### ------- PROFILE -------
        scenario name:
          personal name:
            jp: {student["profile"]["PersonalNameJp"]}

          family name:
            jp: {student["profile"]["FamilyNameJp"]} : ruby: {student["profile"]["FamilyNameRubyJp"]}

          full name:
            jp: {student["profile"]["FamilyNameJp"] + " " + student["profile"]["PersonalNameJp"]} : ruby: {student["profile"]["FamilyNameRubyJp"] + " " + student["profile"]["PersonalNameJp"]}
        
        status message:
          jp: {student["profile"]["StatusMessageJp"]}
        
        age: {student["profile"]["CharacterAgeJp"][:2] if student["profile"]["CharacterAgeJp"][:2].isdigit() else "null   # " + ("secrecy" if student["profile"]["CharacterAgeJp"][:2] == "極秘" else student["profile"]["CharacterAgeJp"][:2])}
        height: {student["profile"]["CharHeightJp"][:3]}
        school year: {student["profile"]["SchoolYearJp"][0]}

        birthday:
          day: {student["profile"]["BirthDay"].split("/")[1]}
          month: {student["profile"]["BirthDay"].split("/")[0]}
        
        hobbies:""")

        if len(re.split(", |、", student["profile"]["HobbyJp"], flags=re.UNICODE)) > 0:
            result = result + "\n"
            for hobby in re.split(", |、", student["profile"]["HobbyJp"], flags=re.UNICODE):
                result = result + "  : \"{}\"\n".format(hobby)
        else:
            result = result + ' ""\n'

        result = result + (f"""\nintroduction: [[\n{student["profile"]["ProfileIntroductionJp"]}\n]]\n""")

    return result

def toYamlString(student):
    result = textwrap.dedent(f"""\
    id: {student["Id"]}
    dev name: {student["DevName"]}
    production step: {student["ProductionStep"]} 


    ### ------- INFO -------
    name: {student["StrippedName"]}
    rarity: {student["Rarity"]}{(" - " + '★'*student["DefaultStarGrade"] + '☆'*(student["MaxStarGrade"] - student["DefaultStarGrade"])) if student["MaxStarGrade"] > 0 else ""}

    school: {student["School"]}
    club: "{student["Club"]}"
    tags: [{", ".join('"{}"'.format(tag) for tag in student["Tags"]) if len(student["Tags"]) > 0 else ''}]


    ### ------- STATUS -------
    type: {"striker" if student["SquadType"] == "Main" else "special" if student["SquadType"] == "Support" else student["SquadType"].lower()}
    role: {toEnumStr(student["TacticRole"])}
    position: {toEnumStr(student["TacticRange"])}

    armor: {toEnumStr(student["ArmorType"])}
    bullet: {toEnumStr(student["BulletType"])}

    weapon: {student["WeaponType"] if student["WeaponType"].isupper() else toEnumStr(student["WeaponType"])}
    aim ik: {toEnumStr(student["AimIKType"])}

    equipments: {"[" + ", ".join([equipment.lower() for equipment in student["EquipmentSlot"]]) + "]" if len(student["EquipmentSlot"]) > 0 else "null"}
    """)

    if student["stat"]:
        result = result + textwrap.dedent(f"""

        ### ------- STATS -------
        hp:
          min: {student["stat"]["MaxHP1"]}
          max: {student["stat"]["MaxHP100"]}
        atk:
          min: {student["stat"]["AttackPower1"]}
          max: {student["stat"]["AttackPower100"]}
        def:
          min: {student["stat"]["DefensePower1"]}
          max: {student["stat"]["DefensePower100"]}
        heal:
          min: {student["stat"]["HealPower1"]}
          max: {student["stat"]["HealPower100"]}

        stability:
          rate: {student["stat"]["StabilityRate"]} - {student["stat"]["StabilityRate"] / 100:.0f}%
          point: {student["stat"]["StabilityPoint"]}

        critical:
          crit point: {student["stat"]["CriticalPoint"]}
          crit dmg rate: {student["stat"]["CriticalDamageRate"]} - {student["stat"]["CriticalDamageRate"] / 100:.0f}%

          res point: {student["stat"]["CriticalResistPoint"]}
          res dmg rate: {student["stat"]["CriticalDamageResistRate"]} - {student["stat"]["CriticalDamageResistRate"] / 100:.0f}%
        
        oppression:
          power: {student["stat"]["OppressionPower"]}
          resist: {student["stat"]["OppressionResist"]}

        ammo:
          cost: {student["stat"]["AmmoCost"]}
          count: {student["stat"]["AmmoCount"]}

        range: {student["stat"]["Range"]}
        regen: {student["stat"]["RegenCost"]}
        dodge: {student["stat"]["DodgePoint"]}
        accuracy: {student["stat"]["AccuracyPoint"]}

        environment:
          street: {student["stat"]["StreetBattleAdaptation"]}
          indoor: {student["stat"]["IndoorBattleAdaptation"]}
          outdoor: {student["stat"]["OutdoorBattleAdaptation"]}
        """)
    
    if "profile" in student.keys():
        result = result + textwrap.dedent(f"""

        ### ------- PROFILE -------
        scenario name:
          personal name:
            jp: {student["profile"]["PersonalNameJp"]}

          family name:
            jp: {student["profile"]["FamilyNameJp"]}
            jp ruby: {student["profile"]["FamilyNameRubyJp"]}

          full name:
            jp: {student["profile"]["FamilyNameJp"] + " " + student["profile"]["PersonalNameJp"]}
            jp ruby: {student["profile"]["FamilyNameRubyJp"] + " " + student["profile"]["PersonalNameJp"]}
        
        status message:
          jp: {student["profile"]["StatusMessageJp"]}
        
        age: {student["profile"]["CharacterAgeJp"][:2] if student["profile"]["CharacterAgeJp"][:2].isdigit() else "null   # " + ("secrecy" if student["profile"]["CharacterAgeJp"][:2] == "極秘" else student["profile"]["CharacterAgeJp"][:2])}
        height: {student["profile"]["CharHeightJp"][:3]}
        school year: {student["profile"]["SchoolYearJp"][0]}

        birthday:
          day: {student["profile"]["BirthDay"].split("/")[1]}
          month: {student["profile"]["BirthDay"].split("/")[0]}
        
        hobbies:""")

        if len(re.split(",|、", student["profile"]["HobbyJp"], flags=re.UNICODE)) > 0:
            result = result + "\n"
            for hobby in re.split(",|、", student["profile"]["HobbyJp"], flags=re.UNICODE):
                result = result + "  - \"{}\"\n".format(hobby)
        else:
            result = result + ' ""\n'

        result = result + "\nintroduction: |\n  {}\n".format(student["profile"]["ProfileIntroductionJp"].replace("\n", "\n  ").replace("\n  \n", "\n\n"))

    return result

###==============================================
### Main

characterTable = {}

dataDirPath = "data/ja-JP/Excel"

filePath = os.path.join(dataDirPath, "CharacterExcelTable.json")
with open(filePath, 'r') as f:
    jsonContent = json.load(f)

    for node in jsonContent["DataList"]:
        characterTable[node["Id"]] = node
        characterTable[node["Id"]]["StrippedName"] = stripName(node["DevName"])


filePath = os.path.join(dataDirPath, "CharacterStatExcelTable.json")
with open(filePath, 'r') as f:
    jsonContent = json.load(f)

    for node in jsonContent["DataList"]:
        characterTable[node["CharacterId"]]["stat"] = node


filePath = os.path.join(dataDirPath, "LocalizeCharProfileExcelTable.json")
with open(filePath, 'r') as f:
    jsonContent = json.load(f)

    for node in jsonContent["DataList"]:
        if characterTable[node["CharacterId"]]:
            keys = [
                "StatusMessageJp",
                "FamilyNameJp",
                "FamilyNameRubyJp",
                "PersonalNameJp",
                "SchoolYearJp",
                "CharacterAgeJp",
                "BirthDay",
                "CharHeightJp",
                "HobbyJp",
                "ProfileIntroductionJp"
            ]

            hasAtleastOneKey = False
            for key in keys:
                if node[key]:
                    hasAtleastOneKey = True
                    break

            if hasAtleastOneKey:
                characterTable[node["CharacterId"]]["profile"] = {}

                for key in keys:
                    characterTable[node["CharacterId"]]["profile"][key] = node[key] or ""


###==============================================
resultDirPath = "result"
for fileType in ["zzz", "yaml"]:
    dirPath = os.path.join(resultDirPath, fileType)

    for chara in characterTable.values():
        if chara["TacticEntityType"] == "Student":
            if len(str(chara["Id"])) != 5:
                continue
            elif any(unnecessaryName in chara["DevName"].lower() for unnecessaryName in ["dummy", "tank"]):
                continue

            filePath = os.path.join(dirPath, chara["TacticEntityType"].lower(), chara["StrippedName"].lower() + "." + fileType)
            os.makedirs(os.path.dirname(filePath), exist_ok=True)

            with open(filePath, 'w') as f:
                if fileType == "yaml":
                    f.write(toYamlString(chara))
                elif fileType == "zzz":
                    f.write(toZZZString(chara))
        
        else:
            filePath = os.path.join(dirPath, chara["TacticEntityType"].lower(), chara["StrippedName"].lower() + "." + fileType)
            os.makedirs(os.path.dirname(filePath), exist_ok=True)

            with open(filePath, 'w') as f:
                if fileType == "yaml":
                    f.write(toYamlString(chara))
                elif fileType == "zzz":
                    f.write(toZZZString(chara))

File: ba_translate.nim
import std/[json, tables, strformat, strutils, os]

func getProperName(s: string): string =
    result = s
    result.removeSuffix("_default")
    result = result.replace('_', ' ')
    # wrong names by dev
    result = result.replace("Aris", "Alice")
    result = result.replace("Hihumi", "Hifumi")
    result = result.replace("Zunko", "Junko")

    result = result.capitalizeAscii()

func toEnumStr(s: string): string =
    if s.len == 0: return ""

    block:
        var tempStr = ""
        for i, c in s:
            if i == 0: tempStr.add(c); continue
            if c.isUpperAscii() and s[i-1].isAlphaAscii():
                tempStr.add('_')
            tempStr.add(c)
        
        result = tempStr
    result = result.toLowerAscii()

###==============================================
### `Student` defination
type Student* = ref object
    kind*: string

    id*: uint32
    devName*, name*: string

    rarity*: string
    star*, maxStar*: uint8

    school*: string
    club*: string
    tags*: seq[string]

    tacticRole*, tacticRange*: string
    squad*, armor*, bullet*, weapon*, aimIk*: string
    equipments*: seq[string]


### `Student` functions
func initFromJson(node: JsonNode): Student =
    result = new Student

    result.kind = node["TacticEntityType"].getStr

    result.id      = node["Id"].getInt.uint32
    result.devName = node["DevName"].getStr
    result.name    = getProperName(result.devName)

    result.rarity  = node["Rarity"].getStr
    result.star    = node["DefaultStarGrade"].getInt.uint8
    result.maxStar = node["MaxStarGrade"].getInt.uint8

    result.school = node["School"].getStr
    result.club   = node["Club"].getStr
    for tag in node["Tags"].items:
        result.tags.add tag.getStr

    result.squad  = node["SquadType"].getStr
    result.armor  = node["ArmorType"].getStr
    result.bullet = node["BulletType"].getStr
    result.weapon = node["WeaponType"].getStr
    result.aimIk  = node["AimIKType"].getStr

    result.tacticRange = node["TacticRange"].getStr
    result.tacticRole  = node["TacticRole"].getStr

    for equipment in node["EquipmentSlot"].items:
        result.equipments.add equipment.getStr

func toZzz(student: Student): string =
    result.add dedent fmt"""
    id: {student.id}
    dev name: {student.devName}


    ### ----- Info -----
    name: {student.name}
    rarity: {student.rarity}"""

    if student.star > 0:
        result.add " : "

        # add real star '★'
        for _ in 1'u8 .. student.star:
            result.add "★"
        
        # add placeholder star '☆'
        for _ in 1'u8 .. (student.maxStar - student.star):
            result.add "☆"
    
    result.add "\n\n"
    result.add dedent fmt"""
    school: {student.school}
    club: "{student.club}"
    tags:"""

    if student.tags.len == 0:
        result.add "\"\"\n"
    else:
        for i in 0 ..< student.tags.len-1:
            result.add &" \"{student.tags[i]}\" :;"
        result.add &" \"{student.tags[^1]}\""
        result.add '\n'
    result.add "\n\n"
    
    result.add dedent fmt"""
    ### ----- Status -----
    type: """
    if student.squad == "Main":
        result.add "striker\n"
    elif student.squad == "Support":
        result.add "special\n"
    else: result.add student.squad.toEnumStr & '\n'

    result.add dedent fmt"""
    role: {student.tacticRole.toEnumStr}
    position: {student.tacticRange.toEnumStr}

    armor:  {student.armor.toEnumStr}
    bullet: {student.bullet.toEnumStr}

    weapon: {student.weapon}
    aim ik: {student.aimIk.toEnumStr}

    equipments:"""
    if student.equipments.len > 0:
        for i in 0 ..< student.equipments.len-1:
            result.add &" {student.equipments[i].toEnumStr} :;"
        result.add &" {student.equipments[^1].toEnumStr}"
    result.add "\n\n\n"

    result.add dedent fmt"""
    ### ----- Stats -----
    """

func toYaml(student: Student): string =
    result.add dedent fmt"""
    id: {student.id}
    dev name: {student.devName}


    ### ----- Info -----
    name: {student.name}
    rarity: {student.rarity}"""

    if student.star > 0:
        result.add " - "

        # add real star '★'
        for _ in 1'u8 .. student.star:
            result.add "★"
        
        # add placeholder star '☆'
        for _ in 1'u8 .. (student.maxStar - student.star):
            result.add "☆"
    
    result.add "\n\n"
    result.add dedent fmt"""
    school: {student.school}
    club: "{student.club}"
    tags: ["""

    if student.tags.len == 0:
        result.add "\"\"\n"
    else:
        for i in 0 ..< student.tags.len-1:
            result.add &"\"{student.tags[i]}\", "
        result.add &"\"{student.tags[^1]}\"]"
        result.add '\n'
    result.add "\n\n"
    
    result.add dedent fmt"""
    ### ----- Status -----
    type: """
    if student.squad == "Main":
        result.add "striker\n"
    elif student.squad == "Support":
        result.add "special\n"
    else: result.add student.squad.toEnumStr & '\n'

    result.add dedent fmt"""
    role: {student.tacticRole.toEnumStr}
    position: {student.tacticRange.toEnumStr}

    armor:  {student.armor.toEnumStr}
    bullet: {student.bullet.toEnumStr}

    weapon: {student.weapon}
    aim ik: {student.aimIk.toEnumStr}

    equipments: ["""
    if student.equipments.len > 0:
        for i in 0 ..< student.equipments.len-1:
            result.add &"{student.equipments[i].toEnumStr}, "
        result.add &"{student.equipments[^1].toEnumStr}]"
    result.add "\n\n\n"

    result.add dedent fmt"""
    ### ----- Stats -----
    """

###=============================================
# directory contains all json files
let dirPath = "data/ja-JP/Excel"

# result table
var stundentTable = newTable[uint32, Student]()


# load 'CharacterExcelTable.json'
block:
    let
      fileName = "CharacterExcelTable.json"
      filePath = dirPath / fileName
    
    let jsonContent = parseFile(filePath)["DataList"]
    for node in jsonContent.items:
        block blk:
            # early out if node type is not student
            if node["TacticEntityType"].getStr != "Student": continue
            if node["Id"].getInt.intToStr.len != 5: continue

            for unnecessaryName in ["dummy", "tank"]:
                if node["DevName"].getStr.toLowerAscii().contains(unnecessaryName):
                    break blk

            let student = initFromJson(node)
            stundentTable[student.id] = student

when isMainModule:
    let destDir = "result"

    for node in stundentTable.values:
        block:
            let destDir  = destDir / "zzz"
            createDir(destDir)

            let
              fileName = node.name.toLowerAscii() & ".zzz"
              filePath = destDir / fileName
            writeFile(filePath, node.toZzz())

        block:
            let destDir  = destDir / "yaml"
            createDir(destDir)

            let
              fileName = node.name.toLowerAscii() & ".yaml"
              filePath = destDir / fileName
            writeFile(filePath, node.toYaml())